apps/asap-server/test/utils/google-sync-calendar.test.ts

```diff
@@ -0,0 +1,86 @@
+import { Auth } from 'googleapis';
+
+import { syncCalendarFactory } from '../../src/utils/sync-google-calendar';
+import * as fixtures from '../fixtures/google-events.fixtures';
+
+const mockList = jest.fn();
+jest.mock('googleapis', () => ({
+  google: {
+    auth: {
+      GoogleAuth: jest.fn(),
+    },
+    calendar: () => ({
+      events: {
+        list: mockList,
+      },
+    }),
+  },
+}));
+
+describe('Sync calendar util hook', () => {
+  const syncEvent = jest.fn();
+  const syncToken = '';
+  const syncCalendarHandler = syncCalendarFactory(
+    syncToken,
+    syncEvent,
+    {} as Auth.GoogleAuth,
+  );
+
+  const calendarId = 'google-calendar-id';
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  test('Should throw when get unknown error from google', async () => {
+    mockList.mockRejectedValueOnce(new Error('Google Error'));
+    await expect(syncCalendarHandler(calendarId)).rejects.toThrow(
+      'Google Error',
+    );
+  });
+
+  test('Should trigger full sync when syncToken is invalidated', async () => {
+    mockList.mockRejectedValueOnce({ code: '410' });
+    mockList.mockResolvedValueOnce({ data: fixtures.listEventsResponse });
+
+    const result = await syncCalendarHandler(calendarId);
+
+    expect(syncEvent).toBeCalledTimes(2);
+    expect(result).toStrictEqual('next-sync-token-1');
+  });
+
+  test('Should not throw if google sends null items', async () => {
+    mockList.mockResolvedValueOnce({
+      data: {
+        ...fixtures.listEventsResponse,
+        items: null,
+      },
+    });
+
+    await syncCalendarHandler(calendarId);
+    expect(syncEvent).toBeCalledTimes(0);
+  });
+
+  test('Should not throw when syncEvent fails to update event', async () => {
```

Comment:
well actually.. are we sure we dont want the whole thing to throw when it fails to update a single event? Maybe that way the whole lambda would be retried? Just a question, im not really sure.

While I'm on it - I noticed [you used Promise.all](https://github.com/yldio/asap-hub/pull/653/files#diff-316c33d3a09eac72218fdebd1ac2dc43ec0d0667a3fbd4fb872fabbdc7e919b9R47) for syncing events. You then don't actually test the scenario when you have multiple events of which only one fails but you still want to make sure that all the other ones are saved correctly. In fact it would be difficult to test, i think, unless maybe we some time delay/nextTick magic.

Anyway with promise all its possible for one promise to reject almost immediately and thus the execution of the remaining ones may also be terminated (im not entirely sure but i think that's what the [definition](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) says. Regardless of whether you test it or not - I would go for .allSettled

Reply:
Agreed, while I'm not sure how I'm going to test that without changing the function signature, i'm going to use `allSettled`.